docs/decoupled_models.md

```diff
@@ -0,0 +1,78 @@
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+-->
+
+# Decoupled Backends and Models
+
+Triton can support [backends](https://github.com/triton-inference-server/backend)
+and models that send multiple responses for a request or zero responses
+for a request. A backend may also send responses out-of-order relative
+to the order that the request batches are executed.
+
+Implementing and deploying such decoupled models can be more involved than
+than backends which generate exactly one response per request.
+
+## Developing Decoupled C++ Backend
+
+Read carefully about the [Triton Backend API](https://github.com/triton-inference-server/backend/blob/main/README.md#triton-backend-api),
+[Inference Requests and Responses](https://github.com/triton-inference-server/backend/blob/main/README.md#inference-requests-and-responses)
+and [Decoupled Responses](https://github.com/triton-inference-server/backend/blob/main/README.md#decoupled-responses).
+The [repeat backend](https://github.com/triton-inference-server/repeat_backend)
+demonstrates how the Triton Backend API can be used to implement a decoupled
+backend. However, the example is designed to show the flexibility of the
+Triton API and in no way should be used in production. This example circumvents
+the restriction placed by the [instance count](model_configuration.md#instance-groups)
+and allows multiple requests to be in process even for single instance. In
+real deployment, the backend should not allow the caller thread to return from
+TRITONBACKEND_ModelInstanceExecute until that instance is ready to handle another
+set of requests.
+
+
+## Deploying Decoupled Models
+
+The users need to explicitly set [decoupled model transaction policy](model_configuration.md#decoupled)
+in their provided [model configuration](model_configuration.md) file for the
+model. Triton requires this information to enable special handling required
+for decoupled models. Deploying decoupled models without this configuration
+setting will throw errors at the runtime.
+
+## Running Inference on Decoupled Models
+
+[Inference Protocols and APIs](inference_protocols.md) describes various ways
+a client can communicate and run inference on the server. For decoupled models,
+Triton's HTTP endpoint cannot be used for running inference as it supports
+exactly one response per request. Even standard ModelInfer RPC in the GRPC endpoint
+does not support decoupled responses. In order to run inference on a decoupled
+model, the client must use the bi-directional streaming RPC. See
+[here](https://github.com/triton-inference-server/common/blob/main/protobuf/grpc_service.proto)
+for more details. The [decoupled_test.py](../qa/L0_decoupled/decoupled_test.py) demonstrates
```

Comment:
I think pointing to the tests is not very helpful for a user that is just getting started with decoupled API. Maybe we can just point them to the Python backend decoupled client?

Reply:
The test is simple to follow. The test demonstrates the full flexibility of the API. When a user can release request, return response and send Final flags. I don't want to point the users to python backend examples to look at the client. It might confuse them more.